### PR TITLE
Store Cassandra commitlogs to another volume

### PR DESCRIPTION
**Pull request by mwringe**
Since we are dealing with commit logs during shutdown (https://github.com/openshift/origin-metrics/blob/master/cassandra/cassandra-prestop.sh#L3) we no longer need to store the commit logs in persistent storage.

By moving the commitlogs to the local emptystorage, we can remove some of the burden on the network based file storage. This should decrease the amount of issues been seen where network based filesystems are not able to keep up with writes.

### CODE REVIEW

**Comment by bbguimaraes** [comment]
@mwringe: what happens if these commit logs are lost with the container?

Otherwise LGTM.

**Comment by pweil-** [comment]
Anything to worry about here if the pod stops abnormally after running for some time and doesn't get to drain the logs properly because the pre-stop hook isn't run?

**Comment by mwringe** [comment]
@pweil- if the commit logs are not handled by the pre-stop hook, for example there was a power outage or someone killed the docker container directly. Then any metrics in the commit logs but not yet sent to persistent storage will be lost.

**Comment by jcantrill** [comment]
@mwringe do you know if we have published any guarantees regarding the data storage and data loss?

**Comment by mwringe** [comment]
@jcantrill If we have any guarantee, I have not seen it.

**Comment by pweil-** [comment]
> @pweil- if the commit logs are not handled by the pre-stop hook, for example there was a power outage or someone killed the docker container directly. Then any metrics in the commit logs but not yet sent to persistent storage will be lost.

ok, understood, thanks @mwringe .  So with the PV we would be able to recover them.  Without it the data would be lost. 

@jcantrill let's bring up your question at the Thurs arch call and get some opinions.

**Comment by mwringe** [comment]
@pweil- so we used to store the commit logs directly to the PV, but a few people have run into issues where their network based PV is not fast enough to kept up with the writes. If we can move the commit logs to the pod volume instead of storing it into the PV, which should alleviate the issue where users have a slow PV.

But this is at the expense of losing non persisted metric data if the pod is forcefully killed or a power outage occurs. The commit logs are essentially a cache before they get written into the actual storage, so we only lose what has not been written out yet.

The other way to handle this would be to write the commit log to a mounted local volume. This way if a power outage occurs we would still have the commit logs to use. This is not the easiest thing for users to configure but it might be an option.

But if we are going to be writing to a host volume for commit logs, then maybe we should just be writing all data to a host volume and not deal with a PV at all.

**Comment by pweil-** [comment]
@mwringe  I think we're on the same page :+1:  After my comment I was thinking about the registry use case where if you run `oadm registry` you get ephemeral storage but a user can mount a volume to `/registry` to make it persistent.  Maybe it's just a matter of telling users that if they want to persist the data they need a fast disk and a mount.  Whether that's a host mount or not is an admin choice but having the option seems like a good compromise.

**Comment by mwringe** [comment]
This PR has been updated, essentially if someone wants to configure their pod to add a volume for the commit logs it is now possible. By default it will work as it has before.

**Comment by mwringe** [comment]
@bbguimaraes : could you take a look and approve this or not?

File: docs/cassandra_advance.adoc [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+By default, the Cassandra cluster is a single node cluster. To add a second node with 10Gi of storage, you would need to call the following command:
+
+[options="nowrap"]
+----
+$ oc process hawkular-cassandra-node-pv -v \
+"IMAGE_PREFIX=openshift/origin-,IMAGE_VERSION=devel,PV_SIZE=10Gi,NODE=2"
+----
+
+To deploy more nodes, you would need to just increase the `NODE` value.
```
Comment (jcantrill) [inline docs/cassandra_advance.adoc]: To deploy more nodes, increment the `NODE` value.

File: docs/cassandra_advance.adoc [hunk 2]
```diff
@@ -0,0 +1,45 @@
+# Scaling the Cassandra Components
+
+Since the Cassandra nodes use persistent storage, we cannot currently scale up or down using replication controllers.
+
+In a subsequent release, nominal services will allow replication controllers to handle this type of scaling and these extra steps will not be needed. But for now a few extra steps are required.
+
+In order to scale up a Cassandra cluster, you will need to use the `hawkular-cassandra-node` template.
+
+By default, the Cassandra cluster is a single node cluster. To add a second node with 10Gi of storage, you would need to call the following command:
```
Comment (jcantrill) [inline docs/cassandra_advance.adoc]: Add a second node with 10Gi of storage by executing the following command:

File: docs/cassandra_advance.adoc [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+[NOTE]
+====
+when you add a new node to a Cassandra cluster, the data stored in the cluster will rebalance across the cluster. The same thing will happen when you remove a node from the Cluster. Adding and removing Cassandra nodes can be an expensive operation.
+====
+
+# Moving the Commit Logs to Another Volume
+
+By default we store the Cassandra commit logs and data to the same volume. For some users, their network based filesystem may not be enough to keep up with writes required which can result in errors.
+
+A lot of these writes will done by the commit logs, and it is possible to configure Cassandra instance to use another volume for these writes, such as a node mounted volume.
```
Comment (jcantrill) [inline docs/cassandra_advance.adoc]: Something is amiss here in the wording

File: docs/cassandra_advance.adoc [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+----
+$ oc process hawkular-cassandra-node-pv -v \
+"IMAGE_PREFIX=openshift/origin-,IMAGE_VERSION=devel,PV_SIZE=10Gi,NODE=2"
+----
+
+To deploy more nodes, you would need to just increase the `NODE` value.
+
+[NOTE]
+====
+when you add a new node to a Cassandra cluster, the data stored in the cluster will rebalance across the cluster. The same thing will happen when you remove a node from the Cluster. Adding and removing Cassandra nodes can be an expensive operation.
```
Comment (jcantrill) [inline docs/cassandra_advance.adoc]: ... Adding or removing...

or ..modifying the number of members in the cluster...

File: docs/cassandra_advance.adoc [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+# Moving the Commit Logs to Another Volume
+
+By default we store the Cassandra commit logs and data to the same volume. For some users, their network based filesystem may not be enough to keep up with writes required which can result in errors.
+
+A lot of these writes will done by the commit logs, and it is possible to configure Cassandra instance to use another volume for these writes, such as a node mounted volume.
+
+[NOTE]
+====
+If the Cassandra pod is shutdown cleanly, then the commit logs are processed and they are no longer needed when the pod is restarted. If the pod is not shut down properly (eg forcefully stopped, power outage) then the Cassandra pod can process the commit logs are startup. If you decide to use an `emptydir` pod volume for the commit logs, be aware that this volume will not be available when the pod is restarted and any data in the commit logs which have not been processed will be lost. Using a persisted volume will prevent data loss from occuring in this situation.
```
Comment (jcantrill) [inline docs/cassandra_advance.adoc]: ..can process the commit logs..

is it 'can' or 'will'

**Comment by jcantrill** [comment]
LGTM outside of  changes to the notes.

**Comment by mwringe** [comment]
Docs part updated

**Comment by mwringe** [comment]
[merge]

**Pull request merged**